pkg/vm/pipeline2/types.go

```diff
@@ -79,8 +83,113 @@ import (
 //  -> π(c.city, s.city, revenue)
 type Pipeline struct {
 	// attrs, column list.
-	attrs []string
+	attrs     []string
+	attrTypes []plan.ColDef
 	// orders to be executed
 	instructions vm.Instructions
 	reg          *process.WaitRegister
 }
+
+func (p *Pipeline) getValidColLen() int {
+	var count int = 0
+	for _, attr := range p.attrs {
+		if attr != "" {
+			count++
+		}
+	}
+	return count
+}
+
+func (p *Pipeline) getValidAttrs() []string {
+	attrs := make([]string, 0)
+	for _, attr := range p.attrs {
+		if attr != "" {
+			attrs = append(attrs, attr)
+		}
+	}
+	return attrs
+}
+
+// Fill in the blanks to bat
+func resetBat(bat *batch.Batch, pAttrs []string, pAttrTypes []plan.ColDef, proc *process.Process) *batch.Batch {
+	resVectors := make([]*vector.Vector, len(pAttrs))
+	var index int = 0
+	for i, attr := range pAttrs {
+		if attr == "" {
+			scalarVector := proc.AllocScalarVector(types.PlanTypeToType(*(pAttrTypes[i].Typ)))
+			FillScalarValForVector(scalarVector)
+			scalarVector.Length = 1
```

Comment:
The length cannot be 1, it should be equal to len(bat.Zs)

Reply:
ok, I will try it later